io-sim-classes/src/Control/Monad/Class/MonadSay.hs

```diff
@@ -1,12 +1,26 @@
-module Control.Monad.Class.MonadSay where
+module Control.Monad.Class.MonadSay
+  ( MonadSay (..)
+  ) where
+
+import           Control.Concurrent (MVar, newMVar, putMVar, takeMVar)
+import           System.IO.Unsafe (unsafePerformIO)
 
 import           Control.Monad.State
 
 class Monad m => MonadSay m where
   say :: String -> m ()
 
+-- global lock for 'MonadSay' 'IO' instance
+--
+-- TODO: remove this, in favour of iohk-monitoring-framework
+lock :: MVar ()
+lock = unsafePerformIO (newMVar ())
+
 instance MonadSay IO where
-  say = print
+  say a = do
+    _ <- takeMVar lock
+    print a
+    putMVar lock ()
```

Comment:
Agree we should use trace instead. That said, to make these not overlap, you can use `BS.putStr (show a)` instead, and then it's an atomic write.

Reply:
Nice, I'll do that.